swift/extractor/SwiftExtractor.cpp

```diff
@@ -211,3 +212,40 @@ void codeql::extractSwiftFiles(SwiftExtractorState& state, swift::CompilerInstan
     }
   }
 }
+
+static void cleanupPendingDeclarations(SwiftExtractorState& state) {
+  std::vector<const swift::Decl*> worklist;
+  std::copy(std::begin(state.pendingDeclarations), std::end(state.pendingDeclarations),
+            std::back_inserter(worklist));
+
+  for (auto decl : worklist) {
+    if (state.emittedDeclarations.count(decl)) {
+      state.pendingDeclarations.erase(decl);
+    }
+  }
+}
+
+static void extractLazy(SwiftExtractorState& state, swift::CompilerInstance& compiler) {
+  cleanupPendingDeclarations(state);
+  std::vector<const swift::Decl*> worklist;
+  std::copy(std::begin(state.pendingDeclarations), std::end(state.pendingDeclarations),
+            std::back_inserter(worklist));
+
+  for (auto pending : worklist) {
+    extractDeclarations(state, compiler, *pending->getModuleContext(), nullptr, pending);
+  }
+}
+
+void codeql::extractExtractLazyDeclarations(SwiftExtractorState& state,
+                                            swift::CompilerInstance& compiler) {
+  // Just in case
+  const int upperBound = 100;
+  int iteration = 0;
+  while (!state.pendingDeclarations.empty() && iteration++ < upperBound) {
```

Comment:
cosmetic nit, very subjectively I prefer such loops as `for`, I find them more readable
```suggestion
  for (; !state.pendingDeclarations.empty() && iteration < upperBound; ++iteration) {
```

Reply:
I'd still go for `while` if you don't mind. `for` in my head is about iterating over something, while `while` is to run until some condition is met 😄

Reply:
feel absolutely free. The thing on the loops like these, is that actually they do both (iterating and waiting for a condition), so it all boils down to what one sees as the prominent feature in it :smile: